### ISSUE TITLE
Loss: nan value with Skirt

### ISSUE THREAD

**Issue by opendeeple** [hbertiche/NeuralClothSim#7]
Hello, I have used NeuralClothSim for train my custom Skirt and I am getting loss value is nan.
Can you help me to fix it, please?
Config file:
```json
{
    "experiment": {
        "epochs": 5000,
        "batch_size": 32,
        "learning_rate": 1e-3,
        "temporal_window_size": 0.5,
        "reflect_probability": 0.0,
        "motion_augmentation": 0.0,
        "checkpoint": null
    },
    "body": {
        "model": "smpl_female_neutral",
        "skeleton": "smpl",
        "input_joints": [0, 1, 2, 3, 4, 5, 6, 9, 12, 13, 14, 16, 17, 18, 19, 20, 21, 22, 23]
    },
    "garment": {
        "name": "skirt",
        "blend_weights_smoothing_iterations": 10
    },
    "data": {
        "dataset": "smpl",
        "train": "toy.txt",
        "validation": "toy.txt",
        "test": "toy.txt",
        "fps": 15
    },
    "model": {
        "blend_weights_optimize": false
    },
    "loss": {
        "cloth": {
            "type": "baraff98",
            "stretch": 10.0,
            "shear": 1.0
        },
        "bending": 2e-5,
        "collision_weight": 10.0,
        "collision_threshold": 0.004,
        "density": 0.15,
        "pinning": 1e2,
        "gravity": "-Y"
    }
}
```
Garment image with blender:
![image](https://user-images.githubusercontent.com/74002741/226584221-0c28651f-a6df-4a9a-8cab-85fe186659ec.png)
Log from train script
```log
Epoch 1/5000
4/4 [==============================] - ETA: 0s - m/Loss: nan - m/Stretch: nan - m/Shear: nan - m/Bending: nan - m/Collision: 1.5867e-04 - m/Gravity: nan - m/Inertia: nanWARNING:absl:Found untraced functions such as Collision_layer_call_fn, Collision_layer_call_and_return_conditional_losses while saving (showing 2 of 2). These functions will not be directly callable after loading.
4/4 [==============================] - 11s 3s/step - m/Loss: nan - m/Stretch: nan - m/Shear: nan - m/Bending: nan - m/Collision: 1.5867e-04 - m/Gravity: nan - m/Inertia: nan - val_m/Loss: nan - val_m/Stretch: nan - val_m/Shear: nan - val_m/Bending: nan - val_m/Collision: 0.0000e+00 - val_m/Gravity: nan - val_m/Inertia: nan
```
Thank your!

**Comment by hbertiche** [comment]
Skirts are challenging garments due to the rich and complex dynamics they will show, and also because they do not closely follow the body, because of this, and given the size of your skirt, expect some difficulties in training,

one important thing to change is `blend_weights_smoothing_iterations`,
garment blend weights are assigned by proximity to the body, because of this, skirts present a large discontinuity in the middle, where the left side of the skirt is assigned to the left leg, and the right side to the right leg,
by smoothing blend weights many times, we remove this discontinuity,
in my experiments, I tried with a value of `50` for this parameter, nonetheless, the skirt I used was significantly shorter than yours, you might need to fine-tune this value for your specific case,
using my code, if you zero out the network output (garment deformations), you will see the result of LBS with assigned blend weights only, this might be useful to _debug_ this parameter,

another very important parameter to change is `batch_size`, I set a value of `32` for this parameter to allow people to download and test this repo without large GPUs,
in practice, though, this value should be much bigger, specially for garments that will present complex deformations, such as skirts,
in my experiments, I trained with batch sizes up to `2048`, to do so, I trained on multiple GPUs,
the code is already prepared to do this (see README),

finally, it is likely that at some point you will need to increase the value for `temporal_window_size`, this will make training take longer and be more expensive, so I would recommend doing this once you solve the `nan` issue and have some nice results,

I refer again to README for a brief explanation of each config parameter,
I would also recommend you start with a simpler garment (perhaps a shorter skirt) until you find a configuration that works for you, and than increase garment complexity,

I hope I was helpful,
please do not hesitate to ask further quesitons

**Comment by NathanielB123** [comment]
I know it's been a while since this issue was opened but I had a very similar problem and realised it had a different cause to the ones Hugo mentioned above. The NCS model, by default, only takes into-account 16 bones when doing skinning on the garment (i.e. hand/foot bones are ignored). It turns out that if the LBS weights transferred from the body include weights from bones outside of those 16 then skinning ends up producing NaNs.

You can solve this by either just modifying the garment mesh such that there are no vertices close enough to hands/feet to inherit their bone weights, or you can modify which bones are taken into-account in the config file (`"input_joints"`)

**Comment by minggini** [comment]
I did modified the config file (`"input_joints"`) to exclude hands/feet, but still appears NaN in most of the losses (except for the collision loss). Also tried changing the `"blend_weights_smoothing_iterations"` but doesn't work. Can anyone help?
